[sen6x] Add SEN6x sensor support

mebner86 opened the pull request:
# What does this implement/fix?

Adds support for Sensirion SEN6x environmental sensor nodes (cleanly separated from SEN5x product line). Implements basic functionality for all members of product family (SEN62, SEN63C, SEN65, SEN66, SEN68, and SEN69C) and logic to read the available sensor channels where available: PM, RH, T, NOx, VOC, CO2, and formaldehyde. 

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Developer breaking change (an API change that could break external components)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

- fixes [Issue 2995](https://github.com/esphome/feature-requests/issues/2995)

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- [esphome/esphome-docs#5801](https://github.com/esphome/esphome-docs/pull/5801)

## Test Environment

- [ ] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [x] RP2040
- [ ] BK72xx
- [ ] RTL87xx
- [ ] nRF52840

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
sensor:
  - platform: sen6x
    type: SEN69C
    i2c_id: i2c_bus
    temperature:
      name: Temperature
      accuracy_decimals: 1
    humidity:
      name: Humidity
      accuracy_decimals: 0
    pm_1_0:
      name: PM <1µm Weight concentration
      accuracy_decimals: 1
    pm_2_5:
      name: PM <2.5µm Weight concentration
      accuracy_decimals: 1
    pm_4_0:
      name: PM <4µm Weight concentration
      accuracy_decimals: 1
    pm_10_0:
      name: PM <10µm Weight concentration
      accuracy_decimals: 1
    nox:
      name: NOx
    voc:
      name: VOC
      algorithm_tuning:
        index_offset: 100
        learning_time_offset_hours: 12
        learning_time_gain_hours: 12
        gating_max_duration_minutes: 180
        std_initial: 50
        gain_factor: 230
    co2:
      name: Carbon Dioxide
      ambient_pressure_compensation:
        pressure: 1013
    formaldehyde:
      name: Formaldehyde
    temperature_compensation:
      offset: 0
      normalized_offset_slope: 0
      time_constant: 0
      slot: 0
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

Commit:
copied and renamed from sen5x

Commit:
format

Commit:
added pressure and altitude compensation

Commit:
fixed typos

Commit:
add tests

Commit:
codeowners

Commit:
stack switch

Commit:
get_compilation_time_ref to get_compilation_time

Commit:
format

Commit:
clang

swoboda1337 commented:
Note that https://github.com/esphome/esphome/pull/9254 accomplishes the same thing. Not saying which approach is better at this point, just point it out.

Commit:
tidy up

Commit:
import config keys

Commit:
add boilerplate __init__.py

Commit:
cleared __init__.py

lbibass commented:
Personally, I feel like it makes more sense to have the SEN6x and SEN5x sensors be separate components. It would reduce confusion since the part names don't really match. Having the 6 series sensors grouped with the 5 series is rather nonsensical. I prefer this PR over the other.

swoboda1337 commented on esphome/components/sen6x/automation.h:
```diff
@@ -0,0 +1,21 @@
+#include "sen6x.h"
+
+namespace esphome {
+namespace sen6x {
```
```suggestion
namespace esphome::sen6x {
```

swoboda1337 commented on esphome/components/sen6x/sen6x.cpp:
```diff
@@ -0,0 +1,495 @@
+      //   this->sen6x_type_ = SEN68;
+      // else if (std::string_view(this->product_name_.data()) == "SEN69C")
+      //   this->sen6x_type_ = SEN69C;
+      // ESP_LOGD(TAG, "Product name: %s", this->product_name_.data());
```
Remove this?

mebner86 replied:
coming to think of it, i think it would be better for the sensor type to be auto-detected but to give the user the ability to overwrite it. this can be used to "downgrade" a sensor if needbe. see separate commit

swoboda1337 commented on esphome/components/sen6x/automation.h:
```diff
@@ -0,0 +1,21 @@
+
+ protected:
+  SEN6XComponent *sen6x_;
+};
```
Should use parented

swoboda1337 commented on esphome/components/sen6x/sen6x.h:
```diff
@@ -0,0 +1,164 @@
+    }
+  }
+
+  uint32_t seconds_since_last_store_;
```
This is never set to anything but zero?

swoboda1337 commented on esphome/components/sen6x/sensor.py:
```diff
@@ -0,0 +1,322 @@
+                gain_factor=230,
+            ),
+            cv.Optional(CONF_STORE_BASELINE, default=True): cv.boolean,
+            cv.Optional(CONF_VOC_BASELINE): cv.hex_uint16_t,
```
Is this used?

swoboda1337 commented on esphome/components/sen6x/sensor.py:
```diff
@@ -0,0 +1,322 @@
+    CONF_FORMALDEHYDE: "set_hcho_sensor",
+}
+
+SETTING_MAP = {}
```
Never used?

swoboda1337 commented on esphome/components/sen6x/sensor.py:
```diff
@@ -0,0 +1,322 @@
+@automation.register_action(
+    "sen6x.start_fan_autoclean", StartFanAction, SEN6X_ACTION_SCHEMA
+)
+async def sen64_fan_to_code(config, action_id, template_arg, args):
```
sen64?

swoboda1337 commented on esphome/components/sen6x/sen6x.cpp:
```diff
@@ -0,0 +1,495 @@
+  params[3] = tuning.gating_max_duration_minutes;
+  params[4] = tuning.std_initial;
+  params[5] = tuning.gain_factor;
+  auto result = write_command(i2c_command, params, 6);
```
```suggestion
  auto result = this->write_command(i2c_command, params, 6);
```
missing a few other places

swoboda1337 commented on esphome/components/sen6x/sen6x.h:
```diff
@@ -0,0 +1,164 @@
+  MEASUREMENT_INIT_FAILED,
+  PRODUCT_NAME_FAILED,
+  FIRMWARE_FAILED,
+  UNKNOWN
```
UNKNOWN here collides with line 54

swoboda1337 commented on esphome/components/sen6x/sen6x.h:
```diff
@@ -0,0 +1,164 @@
+  ERRORCODE error_code_;
+  std::array<char, 32> serial_number_;
+  bool initialized_{false};
+  bool store_baseline_;
```
Should initialize these?

Commit:
Refactor sen6x: use C++17 namespace syntax and Parented helper

- Replace nested namespace declarations with C++17 compact syntax (namespace esphome::sen6x)
- Update StartFanAction to use Parented<SEN6XComponent> helper class
- Remove manual constructor and member variable boilerplate in automation.h

Commit:
Fix VOC baseline storage timing in sen6x component

The seconds_since_last_store_ counter was never incremented, preventing
VOC baseline values from ever being stored to flash. This fix adds the
elapsed time (update interval) to the counter on each update() call,
matching the pattern used in the sgp4x component.

Without this fix, the baseline storage condition would always be false
(0 > 10800), causing the sensor to lose its calibration after every reboot.

Commit:
Address PR review feedback: add this-> prefix, fix enum collision, initialize member variables, fix typo, and remove unused code

- Add this-> prefix to all write_command calls for consistency
- Rename ERRORCODE::UNKNOWN to UNKNOWN_ERROR to resolve naming collision with Sen6xType::UNKNOWN
- Initialize all member variables with default values
- Fix typo in function name: sen64_fan_to_code -> sen6x_fan_to_code
- Remove unused SETTING_MAP and associated loop
- Remove unused CONF_VOC_BASELINE import and configuration option

Commit:
Make sen6x type parameter optional with automatic detection

The sen6x component now automatically detects the sensor type from the
product name read from the device. The 'type' parameter is now optional.

Changes:
- Made 'type' parameter optional in CONFIG_SCHEMA
- Added automatic type inference from product name in setup()
- Added warning when user-specified type doesn't match detected type
- Added error handling for unknown product names
- Updated tests to include both explicit type and auto-detect scenarios

Commit:
Fix clang-tidy: add braces to if statements in infer_type_from_product_name_

swoboda1337 commented:
2025.12.6 wont work, if you are using the addon you can install the beta or dev version (may have to enable advanced mode in your HA user profile).

swoboda1337 commented:
Normally the second Wednesday is beta and a release the third Wednesday of every month.

madpipeline commented:
I've updated to esphome version `2026.1.0b1` and I was able to compile this component, but I got this warning:

```
src/esphome/components/sen6x/sen6x.cpp: In lambda function:
src/esphome/components/sen6x/sen6x.cpp:147:59: warning: 'std::string esphome::Application::get_compilation_time()' is deprecated: Use get_build_time_string() instead. Removed in 2026.7.0 [-Wdeprecated-declarations]
  147 |         uint32_t hash = fnv1_hash(App.get_compilation_time() + std::to_string(combined_serial));
      |                                   ~~~~~~~~~~~~~~~~~~~~~~~~^~
In file included from src/esphome/components/sen6x/sen6x.h:6,
                 from src/esphome/components/sen6x/sen6x.cpp:1:
src/esphome/core/application.h:308:15: note: declared here
  308 |   std::string get_compilation_time() {
      |               ^~~~~~~~~~~~~~~~~~~~
```

My board is:

```
esp32:
  variant: ESP32C3
  board: dfrobot_beetle_esp32c3
  framework:
    type: esp-idf
```

Commit:
Replace deprecated get_compilation_time() with get_build_time_string()

Replace the deprecated App.get_compilation_time() call with the new
get_build_time_string() API to resolve compilation warning. The new
API requires a buffer parameter instead of returning a string directly.

AidenShaw2020 commented:
Hi, i tried beta and development version of ESPHome, but compile crash on this error:

```
src/main.cpp: In function 'void setup()':
src/main.cpp:456:62: error: no matching function for call to 'esphome::sen6x::StartFanAction<>::StartFanAction(esphome::sen6x::SEN6XComponent*&)'
  456 |   sen6x_startfanaction_id = new sen6x::StartFanAction<>(sen66);
      |                                                              ^
In file included from src/esphome.h:49,
                 from src/main.cpp:3:
src/esphome/components/sen6x/automation.h:9:32: note: candidate: 'esphome::sen6x::StartFanAction<>::StartFanAction()'
    9 | template<typename... Ts> class StartFanAction : public Action<Ts...>, public Parented<SEN6XComponent> {
      |                                ^~~~~~~~~~~~~~
src/esphome/components/sen6x/automation.h:9:32: note:   candidate expects 0 arguments, 1 provided
src/esphome/components/sen6x/automation.h:9:32: note: candidate: 'constexpr esphome::sen6x::StartFanAction<>::StartFanAction(const esphome::sen6x::StartFanAction<>&)'
src/esphome/components/sen6x/automation.h:9:32: note:   no known conversion for argument 1 from 'esphome::sen6x::SEN6XComponent*' to 'const esphome::sen6x::StartFanAction<>&'
src/esphome/components/sen6x/automation.h:9:32: note: candidate: 'constexpr esphome::sen6x::StartFanAction<>::StartFanAction(esphome::sen6x::StartFanAction<>&&)'
src/esphome/components/sen6x/automation.h:9:32: note:   no known conversion for argument 1 from 'esphome::sen6x::SEN6XComponent*' to 'esphome::sen6x::StartFanAction<>&&'
```

swoboda1337 commented on tests/components/sen6x/common.yaml:
Can just keep the explicit one and cut the number of tests in half (or add the type to the other one). We have 1000s of tests and too many slows things down.

swoboda1337 commented on esphome/components/sen6x/sensor.py:
```diff
@@ -0,0 +1,315 @@
+    cv.Schema(
+        {
+            cv.GenerateID(): cv.declare_id(SEN6XComponent),
+            cv.Optional("type"): cv.one_of(
```
```suggestion
            cv.Optional(CONF_TYPE): cv.one_of(
```

swoboda1337 commented on esphome/components/sen6x/sensor.py:
```diff
@@ -0,0 +1,315 @@
+    await i2c.register_i2c_device(var, config)
+
+    if "type" in config:
+        cg.add(var.set_type(config["type"]))
```
```suggestion
    if CONF_TYPE in config:
        cg.add(var.set_type(config[CONF_TYPE]))
```

swoboda1337 commented on esphome/components/sen6x/sen6x.cpp:
```diff
@@ -0,0 +1,526 @@
+            encode_uint24(this->serial_number_[0], this->serial_number_[1], this->serial_number_[2]);
+        // Hash with compilation time and serial number
+        // This ensures the baseline storage is cleared after OTA
+        // Serial numbers are unique to each sensor, so mulitple sensors can be used without conflict
```
This has two issues:
  1. Build time changes with each compile, so baselines are orphaned and the sensor has to re-calibrate from scratch
  2. Old preference entries are never deleted, accumulating over time until you run out of nvs flash
  
  Why clear it over OTA?

mebner86 replied:
**VOC baseline NVS storage: accumulation issue not fully resolved**

I've updated sen6x to use `get_config_version_hash()` to match the pattern introduced in f5592595b for sen5x, sgp4x, and sgp30.

However, I noticed this approach still has an accumulation problem - it's just less severe than using build_time. NVS entries are orphaned whenever:
- ESPHome version changes (upgrades)
- YAML config changes

Over time, devices that receive regular updates will still accumulate orphaned entries in NVS flash.

A fully-fixed approach would use a stable NVS key (based only on sensor serial number) and store the config version hash *inside* the struct to detect when to invalidate:

```cpp
struct Sen6xBaselines {
  int32_t state0;
  int32_t state1;
  uint32_t config_hash;  // to detect config/version changes
} PACKED;

// Stable key - always the same for this sensor
uint32_t hash = fnv1a_hash_extend(fnv1a_hash("sen6x_voc_baseline"), combined_serial);
this->pref_ = global_preferences->make_preference<Sen6xBaselines>(hash, true);

if (this->pref_.load(&this->voc_baselines_storage_)) {
    if (this->voc_baselines_storage_.config_hash != App.get_config_version_hash()) {
        // Config/version changed - discard old baseline, start fresh
        ESP_LOGI(TAG, "Config changed, discarding old VOC baseline");
        this->voc_baselines_storage_ = {};
    }
}
// Later when saving:
this->voc_baselines_storage_.config_hash = App.get_config_version_hash();
```
This would apply to sen5x, sgp4x, and sgp30 as well.

@dwmw2, @bdraco - thoughts on this? Is the current approach intentional (accepting minor accumulation as a tradeoff for simplicity), or would a fix across these components be welcome?

mebner86 replied:
to your question of why clear it over OTA: the whole idea of dumping the baseline stems from the sen5x component i used as the starting point for this pull request. i think the idea was to preserve the baseline during power outages but dump it after a firmware update. i guess this was done in anticipation of firmware updates happening rarely and to prevent "dragging" an extremely wrong baseline (e.g. steming from a coding error in a prior firmware) over firmware updates where it could take ages to find it's way back to reality.

mebner86 replied:
also @dwmw2

dwmw2 replied:
I agree that invalidating the NVS key less frequently probably makes sense. Before #12425 it was being invalidated for *every* build, which was egregious. I didn't want to go too far in that generic commit though; it was left as an exercise for individual components to make decisions about how they should behave.

swoboda1337 replied:
> A fully-fixed approach would use a stable NVS key (based only on sensor serial number) and store the config version hash inside the struct to detect when to invalidate:

Yes that sounds much better

Commit:
Fix VOC baseline NVS storage to use config version hash

Replace build_time-based hash with config_version_hash for VOC baseline
storage key. This fixes two issues:

1. Build time changes with each compile, causing baselines to be orphaned
   and forcing sensor recalibration from scratch after every firmware update
2. Old preference entries were never deleted, accumulating in NVS flash

Now uses get_config_version_hash() (combining ESPHOME_CONFIG_HASH and
ESPHOME_VERSION), matching the pattern introduced in f5592595b for sen5x,
sgp4x, and sgp30 components. Baselines persist across recompiles but are
invalidated when the config or ESPHome version changes.

Commit:
Use stable NVS key to fully prevent baseline accumulation

Store config version hash inside Sen6xBaselines struct instead of using
it in the NVS key. This fully prevents NVS accumulation:

- NVS key is now stable (based only on sensor serial number)
- Config/version changes are detected by comparing stored config_hash
- When mismatch detected, old baseline is discarded but same NVS entry
  is reused

Previously, each config or ESPHome version change created a new NVS
entry, orphaning the old one and gradually filling up NVS flash.

Commit:
Use CONF_TYPE constant instead of string literal

Commit:
Remove auto-detect tests for sen6x to reduce test count

swoboda1337 commented on esphome/components/sen6x/sen6x.cpp:
```diff
@@ -0,0 +1,531 @@
+  params[5] = tuning.gain_factor;
+  auto result = this->write_command(i2c_command, params, 6);
+  if (!result) {
+    ESP_LOGE(TAG, "Set tuning parameters failed (command=%0xX, err=%d)", i2c_command, this->last_error_);
```
```suggestion
    ESP_LOGE(TAG, "Set tuning parameters failed (command=0x%X, err=%d)", i2c_command, this->last_error_);
```
this seems malformed

swoboda1337 commented on esphome/components/sen6x/sen6x.cpp:
```diff
@@ -0,0 +1,531 @@
+    // Store baselines after defined interval or if the difference between current and stored baseline becomes too
+    // much
+    if (this->store_baseline_) {
+      this->seconds_since_last_store_ += this->get_update_interval() / 1000;
```
What if update interval is less than 1 second? Probably best to store the millis() timestamp of the last write and just subtract & compare.

swoboda1337 commented on esphome/components/sen6x/sen6x.h:
```diff
@@ -0,0 +1,164 @@
+    tuning_params.learning_time_offset_hours = learning_time_offset_hours;
+    tuning_params.learning_time_gain_hours = learning_time_gain_hours;
+    tuning_params.gating_max_duration_minutes = gating_max_duration_minutes;
+    tuning_params.std_initial = 50;
```
nit: magic number, could use a constexpr for this

swoboda1337 commented on esphome/components/sen6x/sen6x.h:
```diff
@@ -0,0 +1,164 @@
+    temperature_compensation_ = temp_comp;
+  }
+  void set_pressure_compensation(uint16_t pressure) { pressure_compensation_ = pressure; }
+  void set_altitude_compensation(uint16_t altitude) { altitude_compensation_ = altitude; }
```
```suggestion
  void set_pressure_compensation(uint16_t pressure) { this->pressure_compensation_ = pressure; }
  void set_altitude_compensation(uint16_t altitude) { this->altitude_compensation_ = altitude; }
```

swoboda1337 replied:
Probably more of these

tuct commented:
@mikelawrence my version has all the things you mentioned implemented plus some: like startup_timeout and auto clean like thr sen54
https://github.com/tuct/esphome-projects/tree/main/components/sen6x

tuct committed:
initial changes

tuct committed:
[sen6x] Remove unused binary sensor dependency and simplify state retrieval

tuct commented:
@mebner86 
if you want i prepared a PR to your branch, here are he main changes:
https://github.com/mebner86/esphome/pull/3/changes/ed33db965e779534c7f5348dbe26ab8bb3289758
and
https://github.com/mebner86/esphome/pull/3/changes/d9428df6a9b2745d8aa8c55deb2a5fd8a4d53e8c
full PR (includes latest dev, had compile issues)
https://github.com/mebner86/esphome/pull/3

mebner86 commented:
@tuct that would be wonderful

tuct committed:
Add isMeasurementRunning method to SEN6XComponent for measurement status check

tuct committed:
Log startup delay during sensor value publishing

tuct committed:
Clamp remaining startup delay log to avoid negative values

tuct committed:
fix auto clean mode

tuct commented:
@mebner86 Hi Martin,
PR is ready for your fork. If you want, you can review and merge.
If you want, I can take over and fix the remaining issues. If this is the case, I would simply suggest asking @swoboda1337 to close the PR, and I would create a new one from my fork. Only if you don't want to fix the remaining stuff yourself.
I can also prep the docu PR to the other repo as soon as I know what we will use.

tuct committed:
Refactor SEN6XComponent: remove unused wait_after_stop_ method and add reset_device method

tuct committed:
Fix setup method in SEN6XComponent to ensure proper stopping of measurements and add startup delay configuration in sensor.py

tuct commented:
@mikelawrence @mebner86 
We definitely should combine our efforts into one version and work together to get this done!
The decision was to use a dedicated component rather than enhance the existing sen5x one.
Therefore, I created a PR to @mebner86  fork/branch to integrate the enhanced functionality of my version.
 
The three of us should agree on which fork/branch to use to create the final version.
The fork owner needs to be able to review and merge the other person's PRs, then update the final PR to ESPHome.
I don't want or need to be this person; I just want to contribute to a fully featured version of the component.

I started my version 2 years ago, when the eval kit was released but only as an external component: https://github.com/tuct/esphome_external_components/commit/6ae1a17c93d2007fcd9638d0e4e64a53ba316e5c
I did not check the ESPHome GitHub before; after 2 years, I decided to finish the work...

What I want is 
* full coverage of the functionality described in the data sheet. Including handling of FW version stuff like some functions only available from x.x.x+
* Ability to start/stop measurement.
* Auto Clean implemented in ESPHome
* auto set ambient pressure like other Sensirion components

@mebner86 @mikelawrence 
Which fork/branch should we use? This would also define who the owner of the PR is in esphome and would require a review and merge of the PR from the other two.
I would be happy if we stick with @mebner86 or Big Mike @mikelawrence takes over!

tuct commented:
I have the sen66 (FW 3.0.0) and eval kit(FW ?.?.?) for testing at home.
What versions do you guys have access to?

tuct committed:
Optimize setup method in SEN6XComponent: reduce timeout durations and remove unused product name and serial number retrieval logic

tuct committed:
Add ambient pressure compensation source support in SEN6XComponent

tuct committed:
Refactor SEN6XComponent: rename isMeasurmentRunning to is_measurement_running and remove redundant reset command in setup method

tuct committed:
Add getter methods for product name, serial number, and firmware version in SEN6XComponent

tuct committed:
Add @tuct as code owner for SEN6X component

tuct commented:
@mebner86 @mikelawrence 
Here is my PR to this PR: https://github.com/mebner86/esphome/pull/5
and the full source: https://github.com/tuct/esphome/tree/sen6x-mebner86/esphome/components/sen6x

tuct committed:
Add temperature compensation action and implementation in SEN6XComponent

tuct committed:
Refactor firmware version handling in SEN6XComponent: separate major and minor version retrieval

tuct committed:
Refactor serial number handling in SEN6XComponent: change from uint8_t array to std::string for better memory management and readability

mebner86 commented:
Gentlemen, agree that joining efforts is the way forward. 
my motivation was to get a minimal viable dedicated sen6x version going based on the existing sen5x code. having gone through the experience, i now realize the shortcomings of the sen5x base (as you know).
i do need some time to look at the PR of @tuct to my branch and at the code of @mikelawrence to wrap jmy head around the problem
i don't have a strong opinion on which PR we should do the work nor do i have any ownership claims. i do have some bandwidth to work on this, but only limited experience in participating in such projects. So if one of you wants to take the lead i'm happy to follow, otherwise i'll proceed to review and integrate the PR of @tuct

tuct committed:
Refactor read command handling in SEN6XComponent: extract logic into set_read_command_and_words function for improved readability and maintainability

tuct committed:
Remove unused command constants in SEN6XComponent for cleaner code

Commit:
Use C++17 compact namespace syntax in sen6x component

Commit:
Improve sensor type auto-detection with user override support

Auto-detect sensor type from product name only when no type is
configured. When user specifies a type in YAML, respect it and warn
if it mismatches the detected product. Fail clearly when type cannot
be determined at all.

mikelawrence committed:
Use SUB_SENSOR for sensor definitions

Commit:
Fix config schema consistency and wire up missing to_code handlers

- Use CONF_FORMALDEHYDE instead of local CONF_HCHO for schema key
- Rename ambient_pressure/sensor_altitude/co2_asc config keys to
  ambient_pressure_compensation/altitude_compensation/
  automatic_self_calibration for consistency with scd4x
- Replace local constants with imports from esphome.const where available
- Remove dead to_code for nested CO2 sub-keys that called non-existent
  C++ methods (set_pressure_compensation, set_altitude_compensation)
- Add missing to_code handlers for ambient_pressure_compensation,
  altitude_compensation, automatic_self_calibration, store_baseline,
  auto_cleaning, and temperature_acceleration
- Update test YAML to match new config key names

Commit:
Add code owners for sen6x component

Commit:
Use stable NVS key to prevent VOC baseline accumulation

Replace build_time-based hash with a stable hash derived from the sensor
serial number. Store config_version_hash inside Sen6xBaselines to detect
config or ESPHome version changes and discard stale baselines while
reusing the same NVS entry.

This was originally fixed in bc181ad7ad but reverted when merging PR #5.
Re-applies the same approach, adapted for string-based serial_number_.
See also: esphome/esphome#12553 (swoboda1337 review comment).

Commit:
Fix clang-tidy warnings in sen6x component

- Use range-based for loop for serial number parsing
- Add braces around single-line if body
- Remove unused lambda captures (this, poll_retries)
- Merge identical SEN63C/SEN65 switch branches

Commit:
Address review feedback for sen6x component

- Fix baseline store timing: replace seconds counter (never incremented)
  with millis() timestamp comparison
- Fix UNKNOWN enum collision: rename ERRORCODE::UNKNOWN to UNKNOWN_ERROR
- Fix malformed format string %0xX to 0x%04X
- Add missing member initializers (error_code_, store_baseline_,
  voc_baselines_storage_, last_baseline_store_ms_)
- Use constexpr for constants (SHORTEST_BASELINE_STORE_INTERVAL_MS,
  MAXIMUM_STORAGE_DIFF, NOX_DEFAULT_STD_INITIAL)
- Switch automation actions to Parented<SEN6XComponent> pattern
  with register_parented, consolidate simple action handlers
- Add this-> prefix on all setter methods for consistency
- Remove unused CONF_VOC_BASELINE from schema and import
- Remove stale SEN54/SEN55 comments from header

Based on review comments from esphome/esphome#12553.

Commit:
Fix VOC baseline store functionality

Port fixes from sen5x PR #13469 to sen6x component:

- Move baseline save from top of update() to after measurement reads,
  eliminating overlapping I2C transactions that caused the old baseline
  store to silently fail
- Store VOC algorithm state as native uint16_t[4] instead of packing
  into two int32_t values
- Remove unreliable MAXIMUM_STORAGE_DIFF comparison; save unconditionally
  when the 2-hour interval elapses
- Change store interval from 3H to 2H per Sensirion datasheet
- Add store_baseline to dump_config() output
- Remove PACKED from baseline struct to fix -Waddress-of-packed-member

Commit:
Restructure sensor.py config schema

Cherry-pick improvements from PR #6 (by @mikelawrence):

- Replace GAS_SENSOR schema with _gas_sensor() function, allowing
  different defaults for VOC (index_offset=100, gating=180min) vs
  NOX (index_offset=1, gating=720min)
- Move CO2 settings (ambient_pressure_compensation, altitude_compensation,
  automatic_self_calibration) under the co2 sensor config
- Remove float_previously_pct backward-compat helper (new component)
- Remove slot from temperature_compensation config (only needed
  dynamically via the set_temperature_compensation action)
- Tighten std_initial validation to int_range(10, 5000)
- Update to_code() to read CO2 settings from nested location

Commit:
Fix inconsistencies in sen6x component

- Fix misleading comment/log mentioning "reset" when stop measurement is called
- Fix start_fan_cleaning() losing was_running state across timing guard,
  which caused measurement to never restart after host-timed auto cleaning
- Add type-based guards for VOC/NOx/HCHO/CO2 in stopped-path publish to
  match running-path behavior
- Unify product name parsing to use range-for style matching serial number
- Remove std_initial from NOx tuning schema (was silently ignored by C++)
- Add missing icon to formaldehyde sensor schema
- Reorder CONFIG_SCHEMA to group all measurement sensors together
- Rename funcName to func_name for PEP 8 compliance

mebner86 commented:
@tuct @mikelawrence I think we've progressed well so far. But I see a certain risk of losing the key goal out of sight, namely to **ship code useful to the community**. Also, I fear that this PR is becoming increasingly **hard to review**. Therefore, I propose to split our work into two PRs as structured below. If you agree, i'd proceed to implement this split.

---

### PR 1 — MVP (~800–900 lines)

Core sensor functionality that is immediately useful out of the box:

- **Basic sensor readings:** PM1.0/2.5/4.0/10.0, Temperature, Humidity, VOC, NOx, CO2, HCHO
- **Device type detection** and multi-variant support (SEN62–SEN69C)
- **I2C setup:** serial number, product name, firmware version
- **`dump_config`** with device info
- **Polling/update loop** (stripped of baseline storage and pressure compensation logic)

### PR 2 — Advanced Features (~700–800 lines)

Purely additive features that build on top of PR 1:

- VOC/NOx algorithm tuning (6 parameters each)
- VOC baseline persistence (NVS storage with config hash validation)
- Temperature compensation (offset, slope, time constant)
- Temperature acceleration mode (k, p, t1, t2)
- CO2 compensation (ambient pressure from sensor, altitude, auto self-calibration)
- Auto-cleaning interval configuration
- All 8 automation actions (fan cleaning, CO2 FRC, SHT heater, start/stop measurement, etc.)
- `automation.h` entirely
- Startup delay configuration

### Why this split works well

1. **PR 1 is self-contained** — a working sensor that reads all values out of the box, which is immediately useful
2. **PR 2 is purely additive** — every feature builds on top of PR 1 without changing its core behavior
3. **Clean review boundary** — reviewers can evaluate I2C protocol handling and sensor parsing without getting bogged down in tuning/calibration details
4. **Focused attention where it matters** — the hardest parts to review (baseline persistence, config hashing, NVS wear protection) are isolated in PR 2

tuct commented:
@mebner86 very good way to chunk the pr up into smaller pieces.
We can also initial exclude the coming soon sensors like the sen69c allowing us to get the first pr even smaller by removing the hcho sensor.

Commit:
Strip sen6x to MVP for initial review

Remove advanced features to keep the PR under the 1000-line limit.
Core sensor readings for all 6 variants (SEN62-SEN69C) are retained.

Deferred to follow-up PR:
- VOC/NOx algorithm tuning
- VOC baseline persistence
- Temperature compensation and acceleration
- CO2 compensation (ambient pressure, altitude, ASC)
- Auto-cleaning interval configuration
- All automation actions (fan cleaning, CO2 FRC, SHT heater, etc.)
- Configurable startup delay

Commit:
Reset sensor at startup for clean state

Replace the data-ready check and stop-measurements sequence with
a full device reset at startup. This ensures the sensor starts
from a known good state regardless of prior commands or unclean
power cycles that may leave subsystems (e.g. SHT) unresponsive.

Commit:
Log product name, serial and firmware in dump_config

These are useful for troubleshooting but were previously only
logged at DEBUG level during setup, which is not visible when
connecting via API after boot.

mikelawrence commented:
This is great first stab at getting the SEN6X into ESPHome. At 650 lines how can the devs resist!

I have tested this latest update on a SEN63C, SEN66 and SEN66. All three work fine.

Here are some thoughts.

- When the config includes a sensor that is not available it is silently ignored. There are no compile errors either.
- `setup()` blocks long enough to cause warnings
``` console
[18:09:34.717][W][component:579]: sen6x.sensor took a long time for an operation (85 ms)
[18:09:34.718][W][component:582]: Components should block for at most 30 ms
[18:09:34.719][W][component:579]: api took a long time for an operation (86 ms)
[18:09:34.720][W][component:582]: Components should block for at most 30 ms
```
- The `update()` function has a `get_register()` call with a 20ms blocking delay. This is a get data ready status command.
- There is a retry mechanism in `update()` based on that data ready status command. I did some analysis to determine how often data is NOT ready with an update rate of 1 second and as far as I can tell it is not that common. In fact I see a few in the very at startup and one or two every 5 minutes. Even at slower update_rates I see data NOT readys occasionally so the sensor might be going busy for a moment. Your results may vary. 
The datasheet says if measurements are NOT ready it will return the previous values so this added complexity may not be all that beneficial. Plus you could eliminate the blocking 20ms `get_register()`

mikelawrence commented on esphome/components/sen6x/sen6x.cpp:
```diff
@@ -0,0 +1,390 @@
+  LOG_I2C_DEVICE(this);
+  ESP_LOGCONFIG(TAG, "  Product: %s", this->product_name_.c_str());
+  ESP_LOGCONFIG(TAG, "  Serial: %s", this->serial_number_.c_str());
+  ESP_LOGCONFIG(TAG, "  Firmware: %u.%u", this->firmware_version_major_, this->firmware_version_minor_);
```
The top few log messages should be merged into one.

mikelawrence replied:
```suggestion
  ESP_LOGCONFIG(TAG,
                "sen6x:\n"
                "  Product: %s\n"
                "  Serial: %s\n"
                "  Firmware: %u.%u\n"
                "  Address: 0x%02X\n",
                this->product_name_.c_str(), this->serial_number_.c_str(), this->firmware_version_major_,
                this->firmware_version_minor_, this->address_);
```

mikelawrence commented on esphome/components/sen6x/sen6x.cpp:
```diff
@@ -0,0 +1,390 @@
+  // the sensor needs 100 ms to enter the idle state
+  this->set_timeout(100, [this]() {
+    // Reset the sensor to ensure a clean state regardless of prior commands or power issues
+    if (!this->write_command(SEN6X_CMD_RESET)) {
+      ESP_LOGE(TAG, "Failed to reset sensor");
+      this->mark_failed();
+      return;
+    }
```
The `"Failed to reset sensor"` is technically an i2c communication error and can be replaced with `ESP_LOG_MSG_COMM_FAIL`. The log message will indicate the line number of the failure so you can still tell exactly what command failed.

mikelawrence replied:
```suggestion
    if (!this->write_command(SEN6X_CMD_RESET)) {
      ESP_LOGE(TAG, ESP_LOG_MSG_COMM_FAIL);
      this->mark_failed();
      return;
    }
```

mikelawrence commented on esphome/components/sen6x/sen6x.cpp:
```diff
@@ -0,0 +1,390 @@
+    this->set_timeout(100, [this]() {
+      uint16_t raw_serial_number[16];
+      if (!this->get_register(SEN6X_CMD_GET_SERIAL_NUMBER, raw_serial_number, 16, 20)) {
+        ESP_LOGE(TAG, "Failed to read serial number");
+        this->error_code_ = SERIAL_NUMBER_IDENTIFICATION_FAILED;
+        this->mark_failed();
+        return;
+      }
```
You can add the same message to `this->mark_failed();`, `dump_config()` would indicate a failure and you could remove `error_code_` all together. You would have to add this to all `mark_failed()`. The benefit is reduced flash usage but log messages are less specific.

mikelawrence replied:
```suggestion
      if (!this->get_register(SEN6X_CMD_GET_SERIAL_NUMBER, raw_serial_number, 16, 20)) {
        ESP_LOGE(TAG, ESP_LOG_MSG_COMM_FAIL);
        this->mark_failed(LOG_STR(ESP_LOG_MSG_COMM_FAIL));
        return;
      }
```

mikelawrence commented on esphome/components/sen6x/sen6x.h:
```diff
@@ -0,0 +1,54 @@
+  PRODUCT_NAME_FAILED,
+  FIRMWARE_FAILED,
+  UNKNOWN_ERROR
+};
```
This could be removed if you add messages to `mark_failed(LOG_STR(ESP_LOG_MSG_COMM_FAIL))` as indicated above.

mikelawrence commented on esphome/components/sen6x/sen6x.cpp:
```diff
@@ -0,0 +1,390 @@
+      ESP_LOGD(TAG, "Firmware version %u.%u", this->firmware_version_major_, this->firmware_version_minor_);
+
+      if (!this->write_command(SEN6X_CMD_START_MEASUREMENTS)) {
+        ESP_LOGE(TAG, "Error starting continuous measurements.");
```
ESPHome Codebase standards ask for no punctuation in log messages.

mikelawrence commented on esphome/components/sen6x/sen6x.cpp:
```diff
@@ -0,0 +1,390 @@
+        ESP_LOGW(TAG, "Unknown setup error!");
+        break;
+    }
+  }
```
This could be removed if you add messages to mark_failed(LOG_STR(ESP_LOG_MSG_COMM_FAIL)) as indicated above.

mikelawrence commented on esphome/components/sen6x/sen6x.cpp:
```diff
@@ -0,0 +1,390 @@
+      if (this->sen6x_type_ == UNKNOWN) {
+        this->sen6x_type_ = inferred_type;
+        if (inferred_type == UNKNOWN) {
+          ESP_LOGE(TAG, "Unable to infer sensor type from product name '%s'. Please specify 'type' in configuration.",
```
Log message could be shorter "Product name unknown '%s'"

mikelawrence reviewed (commented):
Great job on the component! Here are few suggestions. None are functional changes.

Commit:
Consolidate log messages and remove error_code_ tracking

Implement review suggestions from mikelawrence. Replace individual
ESP_LOGCONFIG calls in dump_config with a single merged call. Use
standard ESP_LOG_MSG_COMM_FAIL macro instead of custom error strings
to reduce flash usage. Remove ERRORCODE enum and error_code_ member
as they are no longer needed.

mebner86 commented:
Thanks @mikelawrence for these sensible suggestions to keep polishing while we wait for review!
I've implemented them in both the `sen6x` branch and the `sen6x-full` branch to ease the second PR.

Commit:
Shorten log messages and improve setup logging

Shorten log messages for reduced flash usage. Move product/serial/
firmware logging from dump_config to setup callbacks at INFO level.
Pass ESP_LOG_MSG_COMM_FAIL to mark_failed() for persistent status.

Commit:
Eliminate blocking delays in setup and update

Replace blocking get_register() calls (which internally use delay())
with non-blocking write + set_timeout + read patterns to keep each
scheduler callback well under the 30ms component blocking limit.

In setup(), split three consecutive get_register() calls (~25ms each,
~75ms total) across separate set_timeout(0) callbacks.

In update(), split the data-ready status check from a single blocking
get_register(..., 20) into write_command + set_timeout(20) + read_data.

Addresses feedback from @mikelawrence in PR #12553.

mikelawrence commented on esphome/components/sen6x/sen6x.cpp:
```diff
@@ -0,0 +1,371 @@
+
+void SEN6XComponent::dump_config() {
+  ESP_LOGCONFIG(TAG, "sen6x:");
+  LOG_I2C_DEVICE(this);
```
```suggestion
  ESP_LOGCONFIG(TAG,
                "sen6x:\n"
                "  Product: %s\n"
                "  Serial: %s\n"
                "  Firmware: %u.%u\n"
                "  Address: 0x%02X\n",
                this->product_name_.c_str(), this->serial_number_.c_str(), this->firmware_version_major_,
                this->firmware_version_minor_, this->address_);
```

mikelawrence replied:
Did you mean to delete serial number, product name and firmware? I still think you should combine into one log message.

mebner86 replied:
the issue is that they are all empty as this code is executed before the data are available. it's not called again after that. Note: the only reliable way for me to access the earliest log data is by configuring logger to output to hardware UART0 (pin 21 on my esp32) and to use an external [ttl to usb converter](https://www.adafruit.com/product/954).

mebner86 replied:
as @mikelawrence pointed out to me on discord, dump_config() is also called when loading logs via wifi where the fields _are_ populated and thus are useful. so putting it back...

mikelawrence commented on esphome/components/sen6x/sen6x.cpp:
```diff
@@ -0,0 +1,371 @@
+        return;
+      }
+      this->serial_number_.clear();
+      this->serial_number_.reserve(32);
```
```suggestion
      uint16_t raw_serial_number[8];
      if (!this->get_register(SEN6X_CMD_GET_SERIAL_NUMBER, raw_serial_number, 8, 20)) {
        ESP_LOGE(TAG, ESP_LOG_MSG_COMM_FAIL);
        this->mark_failed(LOG_STR(ESP_LOG_MSG_COMM_FAIL));
        return;
      }
      this->serial_number_.clear();
      this->serial_number_.reserve(16);
```

mikelawrence replied:
As per my comment above, you can reduce the blocking time by reading only the necessary part of the serial number (16 bytes or 8 words). This is improving something that is not causing a warning and may not be necessary.

mebner86 replied:
hm, i guess this is a judgement call. my intuition would be to stick to the datasheet and not over-optimize routines that are only called at startup. also, i'd not exclude the possibility that future product managers make full use of the flexibility provided by the datasheet. so unless review insists, i'd let it be...

mikelawrence commented on esphome/components/sen6x/sen6x.cpp:
```diff
@@ -0,0 +1,371 @@
+          return;
+        }
+
+        this->product_name_.clear();
```
Similar to the serial number you can reduce the product name read by quite a bit. The current longest product name is "SEN63C" which only requires 6 bytes or 3 words.

mikelawrence commented on esphome/components/sen6x/sen6x.cpp:
```diff
@@ -0,0 +1,371 @@
+          return;
+        }
+
+        this->product_name_.clear();
```
```suggestion
        uint16_t raw_product_name[3];
        if (!this->get_register(SEN6X_CMD_GET_PRODUCT_NAME, raw_product_name, 3, 20)) {
          ESP_LOGE(TAG, ESP_LOG_MSG_COMM_FAIL);
          this->mark_failed(LOG_STR(ESP_LOG_MSG_COMM_FAIL));
          return;
        }

        this->product_name_.clear();
        this->serial_number_.reserve(6);
```

mikelawrence replied:
Similar to the serial number you can reduce the product name read by quite a bit. The current longest product name is "SEN63C" which only requires 6 bytes or 3 words.

I added the last line `this->serial_number_.reserve(6);` to be consistent but I'm not sure it's necessary.

mebner86 replied:
same as above

mikelawrence replied:
This was all for nought. I had the wrong I2C frequency. New timings are below and are well below the 30ms max.

mikelawrence reviewed (commented):
The latest changes have removed all warnings and the component is still rock solid!

Commit:
Restore product, serial and firmware in dump_config

dump_config is also called when connecting via wifi to retrieve the log when data likely is available.

mebner86 commented:
hi @tyeth, we've purged the non-essential features from this pull request to make it simpler to review. A second PR with these features is coming right after this one. so the config from top of page is outdated. Try this instead (either set `type` to your sensor or remove it for auto-detection. If your SEN6x doesn't have a Formaldehyde sensor for example, you can remove the respective lines...

```yaml
sensor:
  - platform: sen6x
    type: SEN69C
    address: 0x6B
    temperature:
      name: "Temperature"
      accuracy_decimals: 2
    humidity:
      name: "Humidity"
      accuracy_decimals: 2
    pm_1_0:
      name: "PM <1µm Weight concentration"
      id: pm_1_0
      accuracy_decimals: 2
    pm_2_5:
      name: "PM <2.5µm Weight concentration"
      id: pm_2_5
      accuracy_decimals: 2
    pm_4_0:
      name: "PM <4µm Weight concentration"
      id: pm_4_0
      accuracy_decimals: 2
    pm_10_0:
      name: "PM <10µm Weight concentration"
      id: pm_10_0
      accuracy_decimals: 2
    nox:
      name: "NOX"
    voc:
      name: "VOC"
    co2:
      name: "Carbon Dioxide"
    formaldehyde:
      name: Formaldehyde

swoboda1337 commented on esphome/components/sen6x/sen6x.cpp:
```diff
@@ -0,0 +1,377 @@
+        if (c2 == '\0')
+          break;
+        this->serial_number_.push_back(c2);
+      }
```
Word-to-string decoding is copy-pasted for serial number and product name. sen5x uses `sensirion_convert_to_string_in_place()` from sensirion_common, should do the same or extract a helper

swoboda1337 commented on esphome/components/sen6x/sen6x.cpp:
```diff
@@ -0,0 +1,377 @@
+        if (now < this->startup_stable_after_) {
+          ESP_LOGD(TAG, "Startup delay, ignoring values");
+          this->status_clear_warning();
+          return;
```
`now < this->startup_stable_after_` breaks when millis() wraps after ~49 days. Should track elapsed time or use a bool flag thats set after the 60s window

swoboda1337 commented on esphome/components/sen6x/sen6x.cpp:
```diff
@@ -0,0 +1,377 @@
+          return;
+        }
+
+        if (this->pm_1_0_sensor_ != nullptr)
+          this->pm_1_0_sensor_->publish_state(pm_1_0);
+        if (this->pm_2_5_sensor_ != nullptr)
+          this->pm_2_5_sensor_->publish_state(pm_2_5);
+        if (this->pm_4_0_sensor_ != nullptr)
+          this->pm_4_0_sensor_->publish_state(pm_4_0);
+        if (this->pm_10_0_sensor_ != nullptr)
+          this->pm_10_0_sensor_->publish_state(pm_10_0);
+        if (this->temperature_sensor_ != nullptr)
+          this->temperature_sensor_->publish_state(temperature);
+        if (this->humidity_sensor_ != nullptr)
+          this->humidity_sensor_->publish_state(humidity);
+        if (this->voc_sensor_ != nullptr && voc_index >= 0)
+          this->voc_sensor_->publish_state(voc);
+        if (this->nox_sensor_ != nullptr && nox_index >= 0)
+          this->nox_sensor_->publish_state(nox);
+        if (this->hcho_sensor_ != nullptr && hcho_index >= 0)
+          this->hcho_sensor_->publish_state(hcho);
+        if (this->co2_sensor_ != nullptr && co2_index >= 0)
+          this->co2_sensor_->publish_state(co2);
```
No validation that configured sensors match the detected type. If user configures CO2 on SEN62 the sensor silently never publishes. sen5x validates this in setup and disables unsupported sensors with an error log

mikelawrence commented:
@swoboda1337 `sensirion_convert_to_string_in_place()` is actually in `sen5x`not in `sensirion_common`. I've created this PR #14269 to move it to `sensirion_common` and make it available to this PR if desired.

Commit:
Merge branch 'dev' of https://github.com/esphome/esphome into sen6x

Commit:
Address reviewer feedback: string decoding, millis wrap, sensor validation

- Use sensirion_convert_to_string_in_place() from sensirion_common for
  serial number and product name decoding, removing duplicate manual loops
- Replace millis()-based startup deadline (uint32_t wrap-around bug after
  ~49 days) with a bool flag set via set_timeout()
- Validate configured sensors against detected type in setup() and disable
  unsupported ones with an error log, matching sen5x behaviour

Tested on SEN63C and SEN66.

Commit:
Remove redundant index guards from publish block

Since setup() now nullifies sensor pointers for unsupported types,
the voc_index/nox_index/hcho_index/co2_index >= 0 guards at publish
time are redundant. A non-null pointer is sufficient.

swoboda1337 commented on esphome/components/sen6x/sen6x.cpp:
```diff
@@ -0,0 +1,377 @@
+static const uint16_t SEN6X_CMD_READ_MEASUREMENT_SEN69C = 0x04B5;
+
+static const uint16_t SEN6X_CMD_START_MEASUREMENTS = 0x0021;
+static const uint16_t SEN6X_CMD_RESET = 0xD304;
```
```suggestion
static constexpr uint16_t SEN6X_CMD_GET_DATA_READY_STATUS = 0x0202;
static constexpr uint16_t SEN6X_CMD_GET_FIRMWARE_VERSION = 0xD100;
static constexpr uint16_t SEN6X_CMD_GET_PRODUCT_NAME = 0xD014;
static constexpr uint16_t SEN6X_CMD_GET_SERIAL_NUMBER = 0xD033;

static constexpr uint16_t SEN6X_CMD_READ_MEASUREMENT = 0x0300;  // SEN66 only!
static constexpr uint16_t SEN6X_CMD_READ_MEASUREMENT_SEN62 = 0x04A3;
static constexpr uint16_t SEN6X_CMD_READ_MEASUREMENT_SEN63C = 0x0471;
static constexpr uint16_t SEN6X_CMD_READ_MEASUREMENT_SEN65 = 0x0446;
static constexpr uint16_t SEN6X_CMD_READ_MEASUREMENT_SEN68 = 0x0467;
static constexpr uint16_t SEN6X_CMD_READ_MEASUREMENT_SEN69C = 0x04B5;

static constexpr uint16_t SEN6X_CMD_START_MEASUREMENTS = 0x0021;
static constexpr uint16_t SEN6X_CMD_RESET = 0xD304;
```

swoboda1337 commented on esphome/components/sen6x/sen6x.cpp:
```diff
@@ -0,0 +1,377 @@
+#include "esphome/core/hal.h"
+#include "esphome/core/log.h"
+#include <cmath>
+#include <cinttypes>
```
```suggestion
```

swoboda1337 committed:
Apply suggestions from code review

swoboda1337 committed:
Apply suggestion from @swoboda1337

swoboda1337 reviewed (approved):
LGTM. Docs need to be updated though after the starlight migration.
